File: ubel-node/src/php_runner.js
// php_composer_scanner.js
import fs   from "fs";
import path from "path";

export class PhpComposerScanner {

  static inventoryData = [];

  // ─────────────────────────────
  // PURL
  // ─────────────────────────────
  static _composerPurl(name, version) {
    // Composer package names are vendor/package, already lowercase by convention.
    // PURL spec: pkg:composer/vendor/package@version
    const clean = name.toLowerCase();
    return `pkg:composer/${clean}@${version ?? ""}`;
  }

  // ─────────────────────────────
  // Detect composer project root
  // ─────────────────────────────
  static _isComposerRoot(dir) {
    return (
      fs.existsSync(path.join(dir, "composer.json")) &&
      fs.existsSync(path.join(dir, "vendor"))
    );
  }

  // ─────────────────────────────
  // Read installed packages from
  // vendor/composer/installed.json  (Composer v1 & v2)
  // ─────────────────────────────
  static _readInstalledJson(vendorDir) {
    const installedPath = path.join(vendorDir, "composer", "installed.json");
    if (!fs.existsSync(installedPath)) return [];

    let raw;
    try {
      raw = JSON.parse(fs.readFileSync(installedPath, "utf8"));
    } catch {
      return [];
    }

    // Composer v2 wraps the array under { "packages": [...] }
    // Composer v1 is a bare array
    return Array.isArray(raw) ? raw : (raw.packages ?? []);
  }

  // ─────────────────────────────
  // Normalise a version string
  // strips Composer's leading "v" or "V"
  // ─────────────────────────────
  static _normaliseVersion(v) {
    if (!v) return "";
    return v.replace(/^v/i, "");
  }

  // ─────────────────────────────
  // Extract license from package
  // metadata (field is an array or
  // a plain string depending on version)
  // ─────────────────────────────
  static _extractLicense(pkg) {
    const lic = pkg.license ?? pkg.licence ?? "unknown";
    if (Array.isArray(lic)) return lic.join(" OR ") || "unknown";
    return lic || "unknown";
  }

  // ─────────────────────────────
  // Scan a single composer project
  // ─────────────────────────────
  static _scanProject(projectRoot) {
    const vendorDir = path.join(projectRoot, "vendor");
    const packages  = this._readInstalledJson(vendorDir);
    if (!packages.length) return [];

    // First pass – build name index for dependency resolution
    const nameIndex = new Map();   // lowercase name → { name, version, pkg }

    for (const pkg of packages) {
      const rawName = pkg.name;
      if (!rawName) continue;
      const norm    = rawName.toLowerCase();
      const version = this._normaliseVersion(pkg.version ?? pkg.version_normalized ?? "");
      nameIndex.set(norm, { name: rawName, version, pkg });
    }

    // Second pass – build components
    const components = [];

    for (const [norm, { name, version, pkg }] of nameIndex.entries()) {
      const id      = this._composerPurl(name, version);
      const license = this._extractLicense(pkg);

      // Direct require-list for this package (runtime deps)
      const requireMap = pkg.require ?? {};
      const dependencies = Object.keys(requireMap)
        .map(dep => dep.toLowerCase())
        .filter(dep => dep !== "php" && !dep.startsWith("ext-"))   // skip PHP/ext pseudo-deps
        .map(dep => {
          const resolved = nameIndex.get(dep);
          return resolved
            ? this._composerPurl(resolved.name, resolved.version)
            : this._composerPurl(dep, "");
        });

      // Physical install path inside vendor/
      const installPath = path.join(vendorDir, ...name.split("/"));

      components.push({
        id,
        name: norm,
        version,
        type:         "library",
        license,
        ecosystem:    "composer",
        state:        "undetermined",
        scopes:       [],
        dependencies,
        paths:        [installPath],
        project_root: projectRoot,
        dev:          pkg["dev-requirements"] === true || pkg.dev === true  // set by Composer v2
      });
    }

    return components;
  }

  // ─────────────────────────────
  // Assign scopes from root
  // composer.json  require / require-dev
  // ─────────────────────────────
  static _assignScopes(inventory) {
    const byId    = new Map(inventory.map(c => [c.id, c]));
    const nameIdx = new Map();

    for (const comp of inventory) {
      if (!Array.isArray(comp.scopes)) comp.scopes = [];
      const key = comp.name;
      if (!nameIdx.has(key)) nameIdx.set(key, []);
      nameIdx.get(key).push(comp);
    }

    // Group by project root
    const projectGroups = new Map();
    for (const comp of inventory) {
      const root = comp.project_root;
      if (!projectGroups.has(root)) projectGroups.set(root, []);
      projectGroups.get(root).push(comp);
    }

    for (const [projectRoot, comps] of projectGroups.entries()) {
      let rootManifest = {};
      try {
        rootManifest = JSON.parse(
          fs.readFileSync(path.join(projectRoot, "composer.json"), "utf8")
        );
      } catch { /* no manifest – fall back below */ }

      const prod = new Set(
        Object.keys(rootManifest.require ?? {}).map(k => k.toLowerCase())
      );
      const dev  = new Set(
        Object.keys(rootManifest["require-dev"] ?? {}).map(k => k.toLowerCase())
      );

      // Also honour the per-package `dev` flag written by Composer v2
      for (const comp of comps) {
        if (comp.dev) dev.add(comp.name);
      }

      const propagate = (names, scope) => {
        const queue = [];

        for (const n of names) {
          for (const c of (nameIdx.get(n) ?? [])) {
            if (c.project_root === projectRoot) queue.push(c);
          }
        }

        const visited = new Set();

        while (queue.length) {
          const c = queue.shift();
          if (visited.has(c.id)) continue;
          visited.add(c.id);

          if (!c.scopes.includes(scope)) c.scopes.push(scope);

          for (const dep of c.dependencies) {
            const d = byId.get(dep);
            if (d && d.project_root === projectRoot) queue.push(d);
          }
        }
      };

      propagate(prod, "prod");
      propagate(dev,  "dev");

      // Fallback – no manifest or empty require sections
      if (prod.size === 0 && dev.size === 0) {
        for (const c of comps) {
          if (c.scopes.length === 0) c.scopes.push("prod");
        }
      }
    }
  }

  // ─────────────────────────────
  // Merge duplicates by PURL
  // ─────────────────────────────
  static mergeInventoryByPurl(components) {
    const map = new Map();

    for (const comp of components) {
      if (!map.has(comp.id)) {
        map.set(comp.id, { ...comp, paths: [...comp.paths] });
        continue;
      }

      const existing = map.get(comp.id);

      for (const p of comp.paths) {
        if (!existing.paths.includes(p)) existing.paths.push(p);
      }

      for (const s of comp.scopes) {
        if (!existing.scopes.includes(s)) existing.scopes.push(s);
      }
    }

    return [...map.values()];
  }

  // ─────────────────────────────
  // ENTRY
  // ─────────────────────────────
  static async getInstalled(startDir = process.cwd()) {

    this.inventoryData = [];

    const visited = new Set();
    const raw     = [];

    function walk(dir) {
      let entries;
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }

      for (const entry of entries) {
        if (!entry.isDirectory()) continue;
        if (["node_modules", ".git", ".ubel", "vendor"].includes(entry.name)) continue;

        const full = path.join(dir, entry.name);

        if (PhpComposerScanner._isComposerRoot(full)) {
          const key = path.resolve(full);
          if (!visited.has(key)) {
            visited.add(key);
            raw.push(...PhpComposerScanner._scanProject(full));
          }
          // Don't descend into a detected project root – its vendor/ is
          // already handled. Walk sibling dirs only.
          continue;
        }

        walk(full);
      }

      // Also check the startDir itself
    }

    // Check startDir itself before walking children
    if (PhpComposerScanner._isComposerRoot(startDir)) {
      const key = path.resolve(startDir);
      if (!visited.has(key)) {
        visited.add(key);
        raw.push(...PhpComposerScanner._scanProject(startDir));
      }
    }

    walk(startDir);

    const merged = this.mergeInventoryByPurl(raw);
    this._assignScopes(merged);

    // Strip internal-only flag before exposing
    for (const c of merged) delete c.dev;

    this.inventoryData = merged;

    return merged.map(c => c.id);
  }
}

export default PhpComposerScanner;

File: python/ubel/php_runner.py
"""
php_scanner.py — PHP / Composer scanner (getInstalled only).

Mirrors the JS PhpComposerScanner:
  - Detects Composer roots (composer.json + vendor/)
  - Reads vendor/composer/installed.json (v1 bare array, v2 { packages: [...] })
  - Assigns prod/dev scopes via BFS from root composer.json require / require-dev
  - Respects per-package "dev" flag written by Composer v2
  - Merges duplicate PURLs
  - PURL: pkg:composer/<vendor>/<package>@<version>
"""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any, Dict, List, Set


class PhpComposerScanner:

    # ------------------------------------------------------------------ #
    # PURL                                                                 #
    # ------------------------------------------------------------------ #

    
    def _composer_purl(self,name: str, version: str) -> str:
        clean = name.lower()
        return f"pkg:composer/{clean}@{version or ''}"

    # ------------------------------------------------------------------ #
    # Detect Composer root                                                 #
    # ------------------------------------------------------------------ #

    
    def _is_composer_root(self,directory: str) -> bool:
        return (
            os.path.exists(os.path.join(directory, "composer.json")) and
            os.path.exists(os.path.join(directory, "vendor"))
        )

    # ------------------------------------------------------------------ #
    # Read installed.json                                                  #
    # ------------------------------------------------------------------ #

    
    def _read_installed_json(self,vendor_dir: str) -> List[Dict[str, Any]]:
        installed_path = os.path.join(vendor_dir, "composer", "installed.json")
        if not os.path.exists(installed_path):
            return []
        try:
            raw = json.loads(Path(installed_path).read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return []
        # v2 wraps under { "packages": [...] }; v1 is a bare array
        return raw if isinstance(raw, list) else raw.get("packages", [])

    # ------------------------------------------------------------------ #
    # Normalize version                                                    #
    # ------------------------------------------------------------------ #

    
    def _normalise_version(self,v: str) -> str:
        if not v:
            return ""
        return v.lstrip("vV")

    # ------------------------------------------------------------------ #
    # Extract license                                                      #
    # ------------------------------------------------------------------ #

    
    def _extract_license(self,pkg: Dict[str, Any]) -> str:
        lic = pkg.get("license") or pkg.get("licence") or "unknown"
        if isinstance(lic, list):
            return " OR ".join(lic) or "unknown"
        return lic or "unknown"

    # ------------------------------------------------------------------ #
    # Scan one Composer project                                            #
    # ------------------------------------------------------------------ #

    
    def _scan_project(self,project_root: str) -> List[Dict[str, Any]]:
        vendor_dir = os.path.join(project_root, "vendor")
        packages   = self._read_installed_json(vendor_dir)
        if not packages:
            return []

        # Pass 1 – build name index
        name_index: Dict[str, Any] = {}
        for pkg in packages:
            raw_name = pkg.get("name")
            if not raw_name:
                continue
            norm    = raw_name.lower()
            version = self._normalise_version(
                pkg.get("version") or pkg.get("version_normalized") or ""
            )
            name_index[norm] = {"name": raw_name, "version": version, "pkg": pkg}

        # Pass 2 – build components
        components: List[Dict[str, Any]] = []

        for norm, entry in name_index.items():
            name    = entry["name"]
            version = entry["version"]
            pkg     = entry["pkg"]
            cid     = self._composer_purl(name, version)
            license_ = self._extract_license(pkg)

            require_map = pkg.get("require") or {}
            dependencies: List[str] = []
            for dep in require_map:
                dep_lower = dep.lower()
                if dep_lower == "php" or dep_lower.startswith("ext-"):
                    continue
                resolved = name_index.get(dep_lower)
                if resolved:
                    dependencies.append(
                        self._composer_purl(resolved["name"], resolved["version"])
                    )
                else:
                    dependencies.append(self._composer_purl(dep_lower, ""))

            install_path = os.path.join(vendor_dir, *name.split("/"))

            is_dev = (
                pkg.get("dev-requirements") is True or
                pkg.get("dev") is True
            )

            components.append({
                "id":           cid,
                "name":         norm,
                "version":      version,
                "type":         "library",
                "license":      license_,
                "ecosystem":    "php",
                "state":        "undetermined",
                "scopes":       [],
                "dependencies": dependencies,
                "paths":        [install_path],
                "project_root": project_root,
                "dev":          is_dev,
            })

        return components

    # ------------------------------------------------------------------ #
    # Assign scopes                                                        #
    # ------------------------------------------------------------------ #

    
    def _assign_scopes(self,inventory: List[Dict[str, Any]]) -> None:
        by_id: Dict[str, Dict] = {c["id"]: c for c in inventory}
        name_idx: Dict[str, List[Dict]] = {}

        for comp in inventory:
            if not isinstance(comp.get("scopes"), list):
                comp["scopes"] = []
            name_idx.setdefault(comp["name"], []).append(comp)

        project_groups: Dict[str, List[Dict]] = {}
        for comp in inventory:
            project_groups.setdefault(comp["project_root"], []).append(comp)

        for project_root, comps in project_groups.items():
            try:
                root_manifest = json.loads(
                    Path(os.path.join(project_root, "composer.json")).read_text(encoding="utf-8")
                )
            except (OSError, json.JSONDecodeError):
                root_manifest = {}

            prod = {k.lower() for k in root_manifest.get("require", {})}
            dev  = {k.lower() for k in root_manifest.get("require-dev", {})}

            # Also honour per-package dev flag from Composer v2
            for comp in comps:
                if comp.get("dev"):
                    dev.add(comp["name"])

            def propagate(names: Set[str], scope: str) -> None:
                queue = [
                    c for n in names
                    for c in name_idx.get(n, [])
                    if c["project_root"] == project_root
                ]
                visited: Set[str] = set()
                while queue:
                    c = queue.pop(0)
                    if c["id"] in visited:
                        continue
                    visited.add(c["id"])
                    if scope not in c["scopes"]:
                        c["scopes"].append(scope)
                    for dep_id in c["dependencies"]:
                        d = by_id.get(dep_id)
                        if d and d.get("project_root") == project_root:
                            queue.append(d)

            propagate(prod, "prod")
            propagate(dev,  "dev")

            if not prod and not dev:
                for c in comps:
                    if not c["scopes"]:
                        c["scopes"].append("prod")

    # ------------------------------------------------------------------ #
    # Merge by PURL                                                        #
    # ------------------------------------------------------------------ #

    
    def merge_inventory_by_purl(self,components: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        merged: Dict[str, Dict] = {}
        for comp in components:
            cid = comp["id"]
            if cid not in merged:
                clone = dict(comp)
                clone["paths"]  = list(comp.get("paths", []))
                clone["scopes"] = list(comp.get("scopes", []))
                merged[cid] = clone
                continue
            existing = merged[cid]
            for p in comp.get("paths", []):
                if p not in existing["paths"]:
                    existing["paths"].append(p)
            for s in comp.get("scopes", []):
                if s not in existing["scopes"]:
                    existing["scopes"].append(s)
        return list(merged.values())

    # ------------------------------------------------------------------ #
    # ENTRY                                                                #
    # ------------------------------------------------------------------ #

    def get_installed(self, start_dir: str = ".") -> List[str]:
        self.inventory_data = []
        start_dir = os.path.abspath(start_dir)

        visited: Set[str] = set()
        raw: List[Dict[str, Any]] = []

        skip = {"node_modules", ".git", ".ubel", "vendor"}

        def walk(directory: str) -> None:
            try:
                entries = os.scandir(directory)
            except OSError:
                return
            with entries as it:
                for entry in it:
                    if not entry.is_dir(follow_symlinks=False):
                        continue
                    if entry.name in skip:
                        continue
                    full = entry.path
                    if self._is_composer_root(full):
                        key = os.path.realpath(full)
                        if key not in visited:
                            visited.add(key)
                            raw.extend(self._scan_project(full))
                        continue
                    walk(full)

        if self._is_composer_root(start_dir):
            key = os.path.realpath(start_dir)
            if key not in visited:
                visited.add(key)
                raw.extend(self._scan_project(start_dir))

        walk(start_dir)

        merged = self.merge_inventory_by_purl(raw)
        self._assign_scopes(merged)

        for c in merged:
            c.pop("dev", None)

        self.inventory_data = merged
        return [c["id"] for c in merged]